model/publishing/delivery/tasks/RemoteDeliveryPublishingTask.php

```diff
@@ -0,0 +1,290 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2020 (original work) Open Assessment Technologies SA ;
+ */
+declare(strict_types=1);
+
+namespace oat\taoPublishing\model\publishing\delivery\tasks;
+
+use Exception;
+use common_report_Report as Report;
+use core_kernel_classes_Resource;
+use common_exception_Error;
+use common_exception_MissingParameter;
+use GuzzleHttp\Exception\ClientException;
+use GuzzleHttp\Exception\ConnectException;
+use GuzzleHttp\Psr7\MultipartStream;
+use GuzzleHttp\Psr7\Request;
+use League\Flysystem\FileNotFoundException;
+use oat\generis\model\OntologyAwareTrait;
+use oat\oatbox\action\Action;
+use oat\oatbox\filesystem\File;
+use oat\oatbox\filesystem\FileSystemService;
+use oat\oatbox\log\LoggerService;
+use oat\tao\model\taskQueue\QueueDispatcherInterface;
+use oat\tao\model\taskQueue\Task\ChildTaskAwareInterface;
+use oat\tao\model\taskQueue\Task\ChildTaskAwareTrait;
+use oat\tao\model\taskQueue\Task\TaskAwareInterface;
+use oat\tao\model\taskQueue\Task\TaskAwareTrait;
+use oat\taoDeliveryRdf\controller\RestTest;
+use oat\taoDeliveryRdf\model\DeliveryAssemblyService;
+use oat\taoPublishing\model\PlatformService;
+use oat\taoPublishing\model\publishing\delivery\PublishingDeliveryService;
+use oat\taoPublishing\model\publishing\exception\PublishingFailedException;
+use oat\taoPublishing\model\publishing\test\TestBackupService;
+use Psr\Http\Message\ResponseInterface;
+use Zend\ServiceManager\ServiceLocatorAwareInterface;
+use Zend\ServiceManager\ServiceLocatorAwareTrait;
+
+class RemoteDeliveryPublishingTask implements Action, ServiceLocatorAwareInterface, ChildTaskAwareInterface, TaskAwareInterface
+{
+    use ServiceLocatorAwareTrait;
+    use OntologyAwareTrait;
+    use TaskAwareTrait;
+    use ChildTaskAwareTrait;
+
+    /** @var core_kernel_classes_Resource */
+    private $test;
+
+    /** @var core_kernel_classes_Resource */
+    private $delivery;
+
+    /** @var core_kernel_classes_Resource */
+    private $environment;
+
+    /**
+     * @param array  $params
+     * @return Report
+     * @throws common_exception_Error
+     * @throws common_exception_MissingParameter
+     */
+    public function __invoke($params) {
+
+        if (count($params) != 3) {
+            throw new common_exception_MissingParameter();
+        }
+        $this->instantiateInputResources($params);
+
+        return $this->compileTest();
+    }
+
+    /**
+     * @param $params
+     */
+    private function instantiateInputResources($params): void
+    {
+        list($testId, $environmentId, $deliveryId) = $params;
+        $this->test = $this->getResource($testId);
```

Comment:
Don't we need to check that resource exists? probably we can do this with setters
```
$this->setTest($testId)

...
function setTest(string $testId): void
{
  $this->test = this->getResource($testId)
}

function getResource($id): Resource
{
// resource = $this->getResource($id)
// resource->exists
// return resource
// exception if doesn't exist?
}
```

Reply:
I added this validation in the method which does actual remote publishing.

Reply:
but you can define this properties as resources and use strict types everywhere

Reply:
Changed implementation a bit. Resources are still instantiated in the task, but validation is implemented in a service, before resource is used.